node/wrapper-allocator/src/lib.rs

```diff
@@ -0,0 +1,78 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use core::sync::atomic::{ Ordering::SeqCst, AtomicUsize };
+use core::alloc::{GlobalAlloc, Layout};
+use tikv_jemallocator::Jemalloc;
+
+pub struct WrapperAllocatorData {
+    allocated: AtomicUsize,
+    checkpoint: AtomicUsize,
+    peak: AtomicUsize,
+    // limit: AtomicUsize, // Should we introduce a checkpoint limit and fail allocation if the limit is hit?
+}
+
+impl WrapperAllocatorData {
+    /// Marks a new checkpoint. Returns peak allocation, in bytes, since the last checkpoint.
+    pub fn checkpoint(&self) -> usize {
+        let allocated = ALLOCATOR_DATA.allocated.load(SeqCst);
+        let old_cp = ALLOCATOR_DATA.checkpoint.swap(allocated, SeqCst);
+        ALLOCATOR_DATA.peak.swap(allocated, SeqCst).saturating_sub(old_cp)
```

Comment:
While access to each field of `WrapperAllocatorData` is atomic, this method (as others too) loads them one by one, and you can expect anything to happen between these loads. For example:
1. Read `let allocated = ALLOCATOR_DATA.allocated.load(SeqCst);`. Scheduler puts our thread to sleep.
2. Another thread calls `alloc` and then `checkpoint`.
3. We wake up and checkpoint goes back in time due to `allocated` being outdated.

Either these values shouldn't depend on each other and work as plain counters or something, or the whole struct should go under mutex.

I mean, I understand this is our best effort, and thus it depends on what is it we're trying to achieve.

Reply:
> While access to each field of `WrapperAllocatorData` is atomic, this method (as others too) loads them one by one, and you can expect anything to happen between these loads.

Yes, that's exactly the concern I'm worried about! And I'm trying to convince myself it is okay. Not sure at all.

It's okay to have a global lock in the `checkpoint()`. It's definitely not okay to have a global lock in the `alloc()`. I've omitted a mutex in the `checkpoint()` for now as it is supposed to be called from a single thread only, but we can add it there, checkpoint is a rare event. `alloc()` is much more concerning.

Say we have two threads, as in your example, and they're allocating at nearly the same point in time. I'm indexing the values below for clarity.

* Thread1 executes `old_alloc1 = allocated.fetch_add(layout_size1)`. At this point `allocated` is updated and represents the new size.
* Then the context switches to another thread
* Thread2 executes `old_alloc2 = allocated.fetch_add2(layout_size2)`. At this point, `allocated` is updated again and represents the true allocated value. Both `old_alloc1` and `old_alloc2` are set to proper values.
* Thread2 executes `peak.fetch_max(old_alloc2 + layout_size2)`. The peak value is properly updated.
* The context switches back and now thread1 executes `peak.fetch_max(old_alloc1 + layout_size1)`. Here's the tricky part. Although thread1 is not aware of the `allocated` value has already been updated by another thread, it already has proper `old_alloc1`, and the `peak` already stores the proper peak value, including thread2's allocation. What is improper here is the assumption that for thread1 `old_alloc1 + layout_size1` is the new allocated value that should be compared with the peak. But do we care? I believe we don't because thread2 has already updated the peak. I mean, by definition, `old_alloc1 + layout_size1 < old_alloc2  + layout_size2` because `old_alloc2 == old_alloc1 + layout_size1`. So we have no chance of missing the new peak value.

Thus it seems we can avoid a global lock inside the `alloc()` and its siblings, but I'm just afraid of missing something. It's the case where ten eyes are much better than two.

Reply:
Using a spinlock should probably be fine here as the critical section here's going to be very short anyway.

Another more complicated alternative that would help with thread contention would be to make this per thread: put the state in thread local storage and have one spinlock per TLS, and only when grabbing a checkpoint lock them all and collate the data. I don't think it's worth it though.